ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/GSM.hs

```diff
@@ -0,0 +1,698 @@
+              ,
+                GSM.durationUntilTooOld = Just durationUntilTooOld
+              ,
+                GSM.equivalent = (==)   -- unsound, but harmless in this test
+              ,
+                GSM.getChainSyncCandidates = readTVar varCandidates
+              ,
+                GSM.getChainSyncIdlers = readTVar varIdlers
+              ,
+                GSM.getCurrentSelection = readTVar varSelection
+              ,
+                GSM.minCaughtUpDuration = thrashLimit
+              ,
+                GSM.setCaughtUpPersistentMark = \b ->
+                    atomically $ do
+                        writeTVar varMarker $ if b then Present else Absent
+              ,
+                GSM.varLedgerStateJudgement = varJudgment
+              }
+            gsmEntryPoint = case j0 of
+                TooOld      -> GSM.enterOnlyBootstrap gsm
+                YoungEnough -> GSM.enterCaughtUp      gsm
+
+        (hist, _model, res) <- id
+          $ withAsync gsmEntryPoint
+          $ \_async -> runCommands' (pure sm') cmds
+
+        pure
+          $ QC.monadicIO
+          $ QSM.prettyCommands (sm undefined j0) hist
+          $ QSM.checkCommandNames cmds (res QC.=== QSM.Ok)
+
+-----
+
+-- | A block count
+newtype B = B Int
+  deriving stock    (Eq, Ord, Generic, Show)
+  deriving newtype  (Num)
+  deriving anyclass (TD.ToExpr)
+
+-- | A slot count
+newtype S = S Int
+  deriving stock    (Eq, Ord, Generic, Show)
+  deriving newtype  (Num)
+  deriving anyclass (TD.ToExpr)
+
+data UpstreamPeer = Amara | Bao | Cait | Dhani | Eric
+  deriving stock    (Bounded, Enum, Eq, Ord, Generic, Show)
+  deriving anyclass (TD.ToExpr)
+
+-- | The cumulative growth relative to whatever length the initial selection
+-- was and the slot relative to the start of the test (which is assume to be
+-- the exact onset of some slot)
+data Selection = Selection !B !S
+  deriving stock    (Eq, Ord, Generic, Show)
+  deriving anyclass (TD.ToExpr)
+
+-- | The age of the candidate is irrelevant, only its length matters
+newtype Candidate = Candidate B
+  deriving stock    (Eq, Ord, Generic, Show)
+  deriving anyclass (TD.ToExpr)
+
+data MarkerState = Present | Absent
+  deriving stock    (Eq, Ord, Generic, Show)
+  deriving anyclass (TD.ToExpr)
+
+newtype WhetherPrevTimePasses = WhetherPrevTimePasses Bool
+  deriving stock    (Eq, Ord, Generic, Show)
+  deriving anyclass (TD.ToExpr)
+
+data ModelState =
+    ModelTooOld
+  |
+    ModelYoungEnough !SI.Time
+    -- ^ when the model most recently transitioned to 'YoungEnough'
+  deriving stock    (Eq, Ord, Generic, Show)
+  deriving anyclass (TD.ToExpr)
+
+candidateOverSelection ::
+     Selection
+  -> Candidate
+  -> GSM.CandidateVersusSelection
+candidateOverSelection (Selection b _s) (Candidate b') =
+    -- TODO this ignores CandidateDoesNotIntersect, which seems harmless, but
+    -- I'm not quite sure
+    GSM.WhetherCandidateIsBetter (b < b')
```

Comment:
I agree that it seems harmless because a candidate branching off before the immutable tip should only ever be temporary until we disconnect. And by Praos, the best chain in the system does *not* fork off before the immutable tip, so it is not feasible for an adversary to delay us from concluding we are caught up this way.

Another way to think about it: The current GSM logic as well as the model here do not make a difference between `CandidateDoesNotIntersect` and `WhetherCandidateIsBetter True`, so sometimes returning `WhetherCandidateIsBetter True` already gives full code coverage. (More directly, one could also take in something of type `ok` instead of the current type of `candidateOverSelection`.)